designs/2024-config-extends/README.md

```diff
@@ -0,0 +1,534 @@
+
+```js
+
+const extender = new ConfigExtender();
+
+class FlatConfigArray {
+
+    // snip
+
+    normalize(context) {
+
+        // make sure not to make any changes if already normalized
+        if (this.isNormalized()) {
+            throw new Error("...");
+        }
+
+        // replace each element with an array
+        this.forEach((element, i) => {
+            const configs = Array.isArray(element) ? element : [element];
+            this[i] = configs.flat().map(config => extender.evaluate(config))
+        });
+
+        // proceed as usual
+        return super.normalize(context)
+            .catch(error => {
+                if (error.name === "ConfigError") {
+                    throw wrapConfigErrorWithDetails(error, this[originalLength], this[baseLength]);
+                }
+
+                throw error;
+
+            });
+    }
+}
+```
+
+## Documentation
+
+This will require documentation changes and an introductory blog post.
+
+At a minimum, these pages will have to be updated:
+
+* https://eslint.org/docs/latest/use/getting-started#configuration
+* https://eslint.org/docs/latest/use/configure/configuration-files
+* https://eslint.org/docs/latest/extend/plugins#configs-in-plugins
+* https://eslint.org/docs/latest/use/configure/combine-configs
+
+## Drawbacks
+
+1. Introducing a new key to flat config at this point means that users of ESLint v8 won't be able to use it. Even though v8 is EOL, we had made an attempt to provide forward compatibility to make people's transition easier.
+1. While this may reduce complexity for users, it increases the complexity of config evaluation inside of ESLint. This necessarily means a performance cost that we won't be able to quantify until implementation.
+1. Users will have to know which version of ESLint supports `extends` in order to use it. There really isn't an easy way to feature test this capability other than to inspect the version of the current ESLint package.
+
+## Backwards Compatibility Analysis
+
+This proposal is additive and does not affect the way existing configurations are evaluated.
+
+## Alternatives
+
+1. **A utility package.** Instead of including `extends` in flat config itself, we could create a utility package that provides a method to accomplish the same thing. This has the advantage that ESLint v8 users could also use this utility package, but the downside that it would require users to install Yet Another Package to do something that many feel should be available out of the box.
+1. **A utility function.** A variation of the previous approach is to export a function from the `eslint` package that people could use to extend configs. This has the advantage that it would be easier to feature test for but the downside that it still requires an extra step to use `extends`.
```

Comment:
Commenting here for discoverability - this is being discussed in https://github.com/eslint/rfcs/pull/126/files#r1860258048